File: Engine/Tools/WismutCLI/Commands/NewProject.cs
namespace WismutCLI;

public static partial class Commands
{
	[WismutCommand("new-project")]
	public static void NewProject(
		[WismutCommandParam("wismut-path", required: true)] string wismutRoot,
		[WismutCommandParam("template")] string template,
		[WismutCommandParam("location")] string outputPath,
		[WismutCommandParam("name")] string name
	)
	{
		if (string.IsNullOrEmpty(template))
			template = "Default";

		if (string.IsNullOrEmpty(outputPath))
			template = "./";

		var templatePath = Path.Combine(wismutRoot, "Engine", "Templates", template);

		Console.WriteLine($"Selected template: {template}\n" +
		                  $"Creating project... ");

		string relativePathToWismut = Path.TrimEndingDirectorySeparator(Path.GetRelativePath(outputPath, wismutRoot));
		CopyDirectoryContentsAndReplaceArgs(templatePath, outputPath, name, relativePathToWismut);

		Console.WriteLine("Done.");
	}

	private static bool CopyDirectoryContentsAndReplaceArgs(string sourceDir, string destinationDir, string projectName, string pathToWismut)
	{
		var dir = new DirectoryInfo(sourceDir);

		if (!dir.Exists)
		{
			Console.WriteLine($"Can't find source path: {sourceDir}");
			return false;
		}

		Directory.CreateDirectory(destinationDir);

		var files = dir.GetFiles();
		foreach (FileInfo file in files)
		{
			string tempPath = Path.Combine(destinationDir, file.Name);
			string content = File.ReadAllText(file.FullName);
			content = content.Replace("{name}", projectName);
			content = content.Replace("{path_to_wismut}", pathToWismut);
			File.WriteAllText(tempPath, content);
		}

		var subDirs = dir.GetDirectories();
		foreach (DirectoryInfo subDir in subDirs)
		{
			string tempPath = Path.Combine(destinationDir, subDir.Name);
			CopyDirectoryContentsAndReplaceArgs(subDir.FullName, tempPath, projectName, pathToWismut);
		}

		return true;
	}
}

File: Engine/Tools/CLI/Source/Commands/NewProject.cpp
#include "Commands/NewProject.h"
#include "CommandRegistry.h"
#include <iostream>
#include <filesystem>
#include <fstream>

namespace fs = std::filesystem;

namespace Commands
{
	static void CopyDirectoryContentsAndReplaceArgs(
		const fs::path& sourceDir,
		const fs::path& destinationDir,
		const std::string& projectName,
		const std::string& pathToWismut)
	{
		if (!fs::exists(sourceDir))
		{
			std::cout << "Can't find source path: " << sourceDir << "\n";
			return;
		}

		fs::create_directories(destinationDir);

		// Copy files
		for (const auto& entry : fs::directory_iterator(sourceDir))
		{
			if (entry.is_regular_file())
			{
				std::string content;
				std::ifstream file(entry.path());
				if (file.is_open())
				{
					content = std::string((std::istreambuf_iterator<char>(file)),
										std::istreambuf_iterator<char>());
					file.close();
				}

				// Replace placeholders
				size_t pos = 0;
				while ((pos = content.find("{name}", pos)) != std::string::npos)
				{
					content.replace(pos, 6, projectName);
					pos += projectName.length();
				}

				pos = 0;
				while ((pos = content.find("{path_to_wismut}", pos)) != std::string::npos)
				{
					content.replace(pos, 16, pathToWismut);
					pos += pathToWismut.length();
				}

				fs::path destFile = destinationDir / entry.path().filename();
				std::ofstream outFile(destFile);
				if (outFile.is_open())
				{
					outFile << content;
					outFile.close();
				}
			}
		}

		// Recursively copy subdirectories
		for (const auto& entry : fs::directory_iterator(sourceDir))
		{
			if (entry.is_directory())
			{
				fs::path destDir = destinationDir / entry.path().filename();
				CopyDirectoryContentsAndReplaceArgs(entry.path(), destDir, projectName, pathToWismut);
			}
		}
	}

	void NewProject(const std::map<std::string, std::string>& arguments)
	{
		// Get parameters
		auto it = arguments.find("wismut-path");
		if (it == arguments.end() || it->second.empty())
		{
			std::cout << "Parameter wismut-path is required\n";
			return;
		}
		std::string wismutRoot = it->second;

		std::string template_name = "Default";
		it = arguments.find("template");
		if (it != arguments.end() && !it->second.empty())
			template_name = it->second;

		std::string outputPath = "./";
		it = arguments.find("location");
		if (it != arguments.end() && !it->second.empty())
			outputPath = it->second;

		std::string projectName = "MyProject";
		it = arguments.find("name");
		if (it != arguments.end() && !it->second.empty())
			projectName = it->second;

		// Build template path
		fs::path templatePath = fs::path(wismutRoot) / "Engine" / "Templates" / template_name;

		std::cout << "Selected template: " << template_name << "\n"
				  << "Creating project... \n";

		// Calculate relative path to Wismut
		fs::path relativePathToWismut = fs::relative(wismutRoot, outputPath);
		std::string relativePathStr = relativePathToWismut.string();

		CopyDirectoryContentsAndReplaceArgs(templatePath, outputPath, projectName, relativePathStr);

		std::cout << "Done.\n";
	}
}
